### ISSUE TITLE
Crash: TypeError: Cannot read property 'kind' of undefined (at isConstraintPosition)

### ISSUE THREAD

**Issue by calebsander** [Microsoft/TypeScript#26978]
**TypeScript Version:**  3.1.0-dev.20180907

**Search Terms:** compile error strict mode

**Code**:
`test.ts`:
```ts
class Abc {
  def: boolean
  constructor() {
    this.def = true
  }
}
```

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "strict": true
  }
}
```

**Expected behavior:**
Running `tsc` successfully compiles the module.

**Actual behavior:**
`tsc` exits with the following error:
```
node_modules/typescript/lib/tsc.js:68651
                throw e;
                ^

TypeError: Cannot read property 'kind' of undefined
    at isConstraintPosition (node_modules/typescript/lib/tsc.js:36644:27)
    at getConstraintForLocation (node_modules/typescript/lib/tsc.js:36653:25)
    at getInitialOrAssignedType (node_modules/typescript/lib/tsc.js:35754:20)
    at getTypeAtFlowAssignment (node_modules/typescript/lib/tsc.js:36094:71)
    at getTypeAtFlowNode (node_modules/typescript/lib/tsc.js:36028:32)
    at getFlowTypeOfReference (node_modules/typescript/lib/tsc.js:35993:51)
    at isPropertyInitializedInConstructor (node_modules/typescript/lib/tsc.js:44521:28)
    at checkPropertyInitialization (node_modules/typescript/lib/tsc.js:44504:50)
    at checkClassLikeDeclaration (node_modules/typescript/lib/tsc.js:44360:17)
    at checkClassDeclaration (node_modules/typescript/lib/tsc.js:44262:13)
```

The code compiles successfully if the `strict` flag is disabled. Also, the code compiles fine if `def` is instead a `number` or `string` (and assigned the appropriate type of value).

**Comment by j-oliveras** [comment]
The last that works is 3.1.0-dev.20180825:
```
C:\code\test>node_modules\.bin\tsc.cmd -v
Version 3.1.0-dev.20180825

C:\code\test>node_modules\.bin\tsc.cmd -p tsconfig.json
```


The first version that fails is 3.1.0-dev.20180828 (after a weekend):
```
C:\code\test>node_modules\.bin\tsc.cmd -v
Version 3.1.0-dev.20180828

C:\code\test>node_modules\.bin\tsc.cmd -p tsconfig.json
C:\code\test\node_modules\typescript\lib\tsc.js:68212
                throw e;
                ^

TypeError: Cannot read property 'kind' of undefined
    at isConstraintPosition (C:\code\test\node_modules\typescript\lib\tsc.js:36319:27)
    at getConstraintForLocation (C:\code\test\node_modules\typescript\lib\tsc.js:36328:25)
    at getInitialOrAssignedType (C:\code\test\node_modules\typescript\lib\tsc.js:35494:20)
    at getTypeAtFlowAssignment (C:\code\test\node_modules\typescript\lib\tsc.js:35819:71)
    at getTypeAtFlowNode (C:\code\test\node_modules\typescript\lib\tsc.js:35753:32)
    at getFlowTypeOfReference (C:\code\test\node_modules\typescript\lib\tsc.js:35718:51)
    at isPropertyInitializedInConstructor (C:\code\test\node_modules\typescript\lib\tsc.js:44131:28)
    at checkPropertyInitialization (C:\code\test\node_modules\typescript\lib\tsc.js:44114:50)
    at checkClassLikeDeclaration (C:\code\test\node_modules\typescript\lib\tsc.js:43970:17)
    at checkClassDeclaration (C:\code\test\node_modules\typescript\lib\tsc.js:43872:13)
```

**Comment by andy-ms** [comment]
I think this has to do with `--strictPropertyInitialization` and `boolean` being a union -- ran into the same problem with `number | null` (but not with `number | undefined`).

**Comment by sirtrollraptor** [comment]
After encountering the same error and bisecting, 0dbad04c3fa77c4a41c78d559ffd5aa1e1ee4b1c is the commit that caused this bug.

**Comment by jonaskello** [comment]
I get a similar crash in tsc 3.0.3:

```
node_modules/typescript/lib/tsc.js:67654
                throw e;
                ^

TypeError: Cannot read property 'kind' of undefined
    at Object.parenthesizeArrayTypeMember (node_modules/typescript/lib/tsc.js:50293:24)
    at Object.createOptionalTypeNode (node_modules/typescript/lib/tsc.js:47548:24)
    at typeReferenceToTypeNode (node_modules/typescript/lib/tsc.js:26427:44)
    at typeToTypeNodeHelper (node_modules/typescript/lib/tsc.js:26238:28)
    at typeReferenceToTypeNode (node_modules/typescript/lib/tsc.js:26415:43)
    at typeToTypeNodeHelper (node_modules/typescript/lib/tsc.js:26238:28)
    at signatureToSignatureDeclarationHelper (node_modules/typescript/lib/tsc.js:26682:52)
    at createTypeNodesFromResolvedType (node_modules/typescript/lib/tsc.js:26521:43)
    at createTypeNodeFromObjectType (node_modules/typescript/lib/tsc.js:26402:35)
    at createAnonymousTypeNode (node_modules/typescript/lib/tsc.js:26357:42)
```

**Comment by mgol** [comment]
> UPDATE: Seems like this is covered by #26153.

@jonaskello I'm not sure. I still see this error in TS `3.1.0-dev.20180919` which was released today and the fix for #26153 was merged 2 days ago.
